### PR TITLE
feat: Add a neard command to replay chain for an archival node 

### PR DESCRIPTION
**Pull request by tayfunelmas**
We would like to validate that we can replay a chain with minimum amount of data. For this, we need a tool that can do that and make sure we do not break this assumption when cleaning up columns from cold DB. See #11827 for details.

For this, we add a new command `neard replay-archive` that is run from an archival node and replays all the chunks and all the blocks for a given range (most importantly from the beginning). This command is most similar to `apply-range` command (in fact we reuse some functions from there). However, we expect to modify the new command more to make it less and less dependent on the non-critical DB columns (eg. incoming and outgoing receipts, ChunkExtra).

In this PR we submit the initial version of the command. We tested the command for a localnet with locust traffic on it. Then we stopped the network and replayed the chain from the beginning.

The main changes are in `tools/replay-archive/` directory. The other changes are for visibility: to be able to share certain functions between the new tool and other tools and core Chain libraries (eg. for tracking progress, validating blocks and chunks, comparing ChunkExtra, extracting incoming receipts). Some of these functions are from `Chain`; we could create a `Chain` instance and run method of it, but `Chain` is currently designed to be used for an ongoing chain head, so instead, we extracted some of its function as static functions.

### CODE REVIEW

File: tools/replay-archive/src/cli.rs [hunk 1]
```diff
@@ -0,0 +1,392 @@
...
+            near_config.config.cold_store.as_ref(),
+        );
+        let storage = opener.open_in_mode(Mode::ReadOnly).context("Failed to open storage")?;
+        match storage.get_split_store() {
+            Some(store) => Ok(store),
+            None => Err(anyhow!("Failed to get split store for archival node")),
+        }
+    }
+
+    /// Replays the enxt block if any. Returns true if there are still blocks to replay
```
Comment (Trisfald) [inline tools/replay-archive/src/cli.rs:138]:
```suggestion
    /// Replays the next block if any. Returns true if there are still blocks to replay
```

**Comment by Trisfald** [comment]
Nice 👍 
Have you also tested the replayability of blocks that have been GC'd in hot storage?

**Comment by tayfunelmas** [comment]
> Have you also tested the replayability of blocks that have been GC'd in hot storage?

My plan is to write a nayduck test that will first run the chain for a while (until it passes the GC epoch limit) and then stop it and run the replay command on the chain. Similarly to the `view_requests_to_archival_nodes.rs` test but in nayduck.

**Review by pugachAG — APPROVED** [review]
LGTM!

File: tools/replay-archive/src/cli.rs (hunk 1 above)
Comment (tayfunelmas) [reply tools/replay-archive/src/cli.rs:138]: done

File: tools/replay-archive/src/cli.rs [hunk 2]
```diff
@@ -0,0 +1,392 @@
...
+        }
+        self.replay_block(self.next_height)?;
+        self.next_height += 1;
+        Ok(self.next_height <= self.end_height)
+    }
+
+    fn replay_block(&mut self, height: BlockHeight) -> Result<()> {
+        tracing::info!("Replaying block at height {}", self.next_height);
+
+        let block_hash = match self.chain_store.get_block_hash_by_height(height) {
```
[the remark this answers is not in the document]
Comment (tayfunelmas) [reply tools/replay-archive/src/cli.rs:152]: done. thanks!

File: tools/state-viewer/src/util.rs [hunk 3]
```diff
@@ -88,3 +93,99 @@ fn get_last_final_from_height(height: u64, head: &Tip, chain_store: &ChainStore)
...
+) -> anyhow::Result<()> {
+    let height = block.header().height();
+    let block_hash = block.header().hash();
+    let protocol_version = block.header().latest_protocol_version();
+    let new_chunk_extra = resulting_chunk_extra(
+        apply_result,
+        block.chunks()[shard_id as usize].gas_limit(),
+        protocol_version,
+    );
+    println!(
```
[the remark this answers is not in the document]
Comment (tayfunelmas) [reply tools/state-viewer/src/util.rs:167]: It was copied from the original location (apply range comment) and meant to be there.

**Pull request merged**